### PR TITLE
Discord API and Bot for PSS13 (EMAIL NOTIFICATIONS)

### PR DESCRIPTION
**Pull request by ingles98**
Adds a very neaty API for all your in-game to Discord needs!

Comes with a working link/unlink and mailing notification system PLUS a working bot for discord.

Has a lot of potential features in the future!

### CODE REVIEW

File: code/modules/discord_persistence/_hook.dm [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+Discord Bot designer specially for the Persistence needs!
+
+On launch, this is currently only used for mailing notifications and linking accounts.
+
+Any troubles just contact me - Stigma
+*/
+
+
+//GLOBAL DEFINE//
+var/global/datum/discord_api/discord_api = new()
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: GLOBAL_DATUM_INIT(discord_api, /datum/discord_api, new)

File: code/modules/discord_persistence/_hook.dm [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	if(!q.Execute(db))
+		message_admins(src.db.ErrorMsg())
+		return
+
+//The mail proc, which handles sending the mail information to the Bot. The rest is taken care of by it.
+/datum/discord_api/proc/mail(receiver_name, var/datum/computer_file/data/email_message/message)
+	var/receiver_ckey = Retrieve_Record(receiver_name).ckey
+	var/sender_name = message.source
+	var/database/query/g = new("SELECT * FROM [discord_api.usersTable] WHERE ckey = ? AND valid = 1", receiver_ckey)
+	if (g.Execute(discord_api.db) && g.NextRow())
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: Shouldn't these be this object's variables, and therefore only require ...FROM [usersTable] WHERE... and ...g.Execute(db)...

File: code/modules/discord_persistence/_hook.dm [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		src.send_message(msg)
+
+//A broadcast bot, for the broadcasting needs. (This was mainly for testing, probably should have no use at all.)
+/datum/admins/proc/discord_broadcast()
+	set category = "Admin"
+	set name = "Broadcast to Discord"
+	set desc = "VERY EARLY DEV"
+	var/msg = input(usr, "Message:", "Discord") as text|null
+	if (msg)
+		discord_api.send_message("BROADCAST|[msg]")
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: GLOB.discord_api.send_message...

File: code/modules/discord_persistence/_hook.dm [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+/client/verb/linkdiscord()
+	set category = "Special Verbs"
+	set name = "Discord Account - Associate"
+	set desc = "Link your discord account to your BYOND account."
+
+	var/userID = input(usr, "Discord User ID:", "Discord") as text|null
+	if (userID)
+		userID = "\"[userID]\""
+		var/database/query/g = new("SELECT * FROM [discord_api.usersTable] WHERE ckey = ? OR userID = ?", usr.ckey, userID)
+		if (g.Execute(discord_api.db) && g.NextRow())
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: ...FROM [GLOB.discord_api.userTable]...
...(GLOB.discord_api.db)...

File: code/modules/discord_persistence/_hook.dm [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			to_chat(usr, SPAN_WARNING("Could not complete your Discord Account link request. It seems you have already linked this BYOND account OR this discord ID is already linked to another account."))
+		else
+			var/database/query/q = new("INSERT INTO [discord_api.usersTable] VALUES(?,?,0) ", userID, usr.ckey)
+			if(!q.Execute(discord_api.db))
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: ...INTO [GLOB.discord_api.usersTable]...
...(GLOB.discord_api.db)...

File: code/modules/discord_persistence/_hook.dm [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+/client/verb/unlinkdiscord()
+	set category = "Special Verbs"
+	set name = "Discord Account - Disassociate"
+	set desc = "Devalidates the link between your BYOND and Discord account."
+
+	var/database/query/g = new("SELECT * FROM [discord_api.usersTable] WHERE ckey = ?", usr.ckey)
+	if (g.Execute(discord_api.db) && g.NextRow())
+		var/database/query/delete = new("DELETE FROM [discord_api.usersTable] WHERE ckey = ?", usr.ckey)
+		if (delete.Execute(discord_api.db))
```
Comment (Andrew-Fall) [inline code/modules/discord_persistence/_hook.dm]: GLOBS

File: code/modules/modular_computers/NTNet/emails/email_account.dm [hunk 7]
```diff
@@ -54,6 +54,7 @@
 
 /datum/computer_file/data/email_account/proc/receive_mail(var/datum/computer_file/data/email_message/received_message, var/relayed)
 	received_message.set_timestamp()
+	discord_api.mail(src.login, received_message)
```
Comment (Andrew-Fall) [inline code/modules/modular_computers/NTNet/emails/email_account.dm]: GLOB.discord_api.mail...

**Review by Andrew-Fall — CHANGES_REQUESTED** [review]

File: code/modules/discord_persistence/_hook.dm (hunk 2 above)
Comment (ingles98) [reply code/modules/discord_persistence/_hook.dm]: What do u mean? Well i am fairly amateur in Db code, but i do think i need the select cmd

Comment (ingles98) [reply code/modules/discord_persistence/_hook.dm]: Oh lmao i know what u mean now. Couldn't get it through my phone somehow. Yes you're right, i think that happened when i was using find/replace in some cases.

**Pull request merged**